packages/deposit_firebase/lib/src/deposit_firebase_adapter.dart

```diff
@@ -0,0 +1,118 @@
+import 'package:cloud_firestore/cloud_firestore.dart';
+import 'package:deposit/deposit.dart';
+
+extension _QueryDocumentSnapshotX on DocumentSnapshot {
+  Map<String, dynamic> toMap() {
+    final _data = (data() ?? {}) as Map<String, dynamic>;
+
+    return <String, dynamic>{
+      'id': id,
+      ..._data,
+    };
+  }
+}
+
+class FirebaseDepositAdapter extends DepositAdapter<String> {
+  FirebaseDepositAdapter({required this.firestore});
+
+  final FirebaseFirestore firestore;
+
+  @override
+  Future<Map<String, dynamic>> add(
+    String table,
+    String primaryColumn,
+    Map<String, dynamic> data,
+  ) async {
+    final value = await firestore.collection(table).add(data);
+
+    final snapshot = await value.snapshots().first;
+    final returnedData = snapshot.data();
+
+    if (returnedData == null) {
+      throw Exception('Unable to insert data into "$table"');
+    } else {
+      return snapshot.toMap();
+    }
+  }
+
+  @override
+  Future<List<Map<String, dynamic>>> by(String table, String key, value) async {
+    final data = await firestore
+        .collection(table)
+        .where(
+          key,
+          isEqualTo: value,
+        )
+        .get();
+
+    return data.docs.map((doc) => doc.toMap()).toList();
+  }
+
+  @override
+  Future<bool> exists(String table, String primaryColumn, String id) async {
+    if (primaryColumn == 'id') {
+      final doc = await firestore.collection(table).doc(id);
+      final snap = await doc.get();
+      return snap.exists;
+    } else {
+      final data = await firestore
+          .collection(table)
+          .where(primaryColumn, isEqualTo: id)
+          .get();
+
+      return data.docs.isNotEmpty;
+    }
+  }
+
+  @override
+  Future<Map<String, dynamic>> getById(
+    String table,
+    String primaryColumn,
+    String id,
+  ) async {
+    if (primaryColumn != 'id') {
+      throw ArgumentError(
+        'Firebase can only get documents by id using the "id" field',
```

Comment:
Cant u use the `where` method? And return the first element? Considering we do allow for checking existance using Id?

Reply:
That was how I was doing at first, but as far as I can tell no, the id seems to not be part of the attributes. So I think this is a limitation from Firebase adapter

Reply:
I meant that if you have a different `primaryColumn` we can look it up using the `where` filtering method. Just like you did in the `exist` method?

Reply:
Got it, I changed it a little bit